README.md
# mmbop

Dynamic DNS ([RFC 2136](https://tools.ietf.org/html/rfc2136)) provides a mechanism for updating entries in a zone file without needing to edit that file by hand and without needing to restart BIND. For Python users, the [dnspython](http://www.dnspython.org/) module provides a valuable foundation for creating dynamic DNS applications. However there are some things that dnspython and dynamic DNS cannot do - specifically, the creation of new authoritative zones.

BIND does come with a program, rndc, that can control various aspects of the name service, locally or remotely. One of the functions is adding a zone. However the command relies on the fact that the empty zone file already exists. So there is a need for a program to wrap rndc functionality and provide push button capability to add/remove zones. This is the purpose of mmbop.

## Getting Started

Clone the repository and you should be almost done. A (successful) attempt was made to only use modules from the standard library, so there are no Python dependencies when using the command-line (mmbop.py). For the API (mmbop_api.py), [Falcon](https://falconframework.org/) is used and required.

### Prerequisites

A working DNS environment would be useful. In *named.conf* (or associated configuration files), you will need a *controls* stanza that specifies the key that rndc is allowed to use, along with the acl defining the source (in our case, mmbop is run locally, so 127.0.0.1 - aka, localhost - is sufficient).

```
include "/etc/bind/rndc.key";

controls {
        inet * allow { localhost; } keys { "rndc-key"; };
};
```

If you are running BIND version 9.11.0 or later and wish to take advantage of the catalog zone feature (allowing you to automatically update secondary nameservers with the new zones you have created), the appropriate configuration is required on both the primary and secondary server(s).

Primary (for example sake, this has IP 10.1.1.1):
```
        catalog-zones {
                zone "catalog.zone";
        };

zone "catalog.zone" {
        type master;
        file "catalog.zone.db";
        allow-update { key rndc-key; };
};
```

Secondary:
```
        catalog-zones {
                zone "catalog.zone" default-masters { 10.1.1.1; };
        };

zone "catalog.zone" {
        type slave;
        file "/etc/bind/catalog.zone.db";
        masters { 10.1.1.1; };
};
```

See [this nice introduction](https://kb.isc.org/docs/aa-01401) to catalog zones for more information.

**Python and BIND version requirements**

The code was developed using Python 3.6.8 and BIND 9.14.3.
The code was successfully deployed using Python 3.7.4 and BIND 9.11.2rc2

Python 3.6+ is required as mmbop relies on a change in *subprocess* that accepts an *encoding* parameter. This matters for the catalog zone update.
BIND 9.11+ is required to support catalog zones.

### Installing

Assuming now you have at least a working primary DNS server that is accepting a local rndc connection, and you cloned this repository.

**Step 1 : Configure mmbop.ini**

Rename the example file to *mmbop.ini* and edit it so that the required values match your environment.

```
[RNDC]
server: 127.0.0.1
port: 953
keyfile: /etc/bind/rndc.key
path: /usr/sbin/rndc

[DNS]
dns1: ns1.example.com
dns2: ns2.example.com
serial: 1
owner: hostmaster.example.com
namedir: /etc/bind
nameown: bind
namegrp: bind
nameper: 0644

[MMBOP]
protect: example.com|example.net
require: .example.com|.example.net
options: also-notify { 10.10.10.1; };|allow-update { key "myddnskey"; };
catalog: catalog.example
Give the example
```
The .ini file comments explain each option, but the 2 least obvious ones (as they are specific to the mmbop application) are *protect* and *require*.

- protect: A list (entries separated by '|') of exact domain names that mmbop is not allowed to manage (add/remove). Typically used to protect the parent zone, if you are looking to make subdomains only.
- require: A list (entries separated by '|') of end of string matches. For mmbop to manage a domain, it must match at least one of these strings. Typically used to specify the subdomain (note preceeding '.' in example above).

**Step 2 : Verify by running mmbop for server status**

```
# python mmbop.py status
version: BIND 9.14.5-Ubuntu (Stable Release) <id:c2c2b6d>
running on example_server: Linux x86_64 4.15.0-62-generic #69-Ubuntu SMP Wed Sep 4 20:55:53 UTC 2019
boot time: Fri, 22 Nov 2019 21:43:09 GMT
last configured: Tue, 14 Jan 2020 19:09:32 GMT
configuration file: /etc/bind/named.conf
CPUs found: 2
worker threads: 2
UDP listeners per interface: 2
number of zones: 14 (0 automatic)
debug level: 0
xfers running: 0
xfers deferred: 0
soa queries in progress: 0
query logging is OFF
recursive clients: 0/900/1000
tcp clients: 4/150
server is up and running
```
If you see output similar to the example above, congrats you have a working mmbop setup.
If this doesn't work, run with verbose logging (*-v*) for more details on the problem.

**Step 3 : Add a zone**

Given what mmbop is required to do to add a domain, it is necessary to run as a user with write permissions to the directory where the zone files are located.

You can run it as root (sudo), or - from root - you can sudo and run as the same user that owns the BIND service and files. By default, on Ubuntu this is the *bind* user, which is created when you install BIND.

```
# python mmbop.py zoneadd nina.example.com
Add of zone nina.example.com succeeded

# python mmbop.py zonestatus nina.example.com
name: nina.example.com
type: master
files: nina.example.com.db
serial: 1
nodes: 1
last loaded: Tue, 14 Jan 2020 19:07:45 GMT
secure: no
dynamic: yes
frozen: no
reconfigurable via modzone: yes
```

**Step 4 : Remove a zone**

```
# python mmbop.py zonedel nina.example.com
Deletion of zone nina.example.com succeeded

# python mmbop.py zonestatus nina.example.com
rndc: 'zonestatus' failed: not found
no matching zone 'nina.example.com' in any view
```

**Step 5 : Explore**

Run with *-h* or *--help* to see all of the available command-line options

```
$ python mmbop.py --help
usage: mmbop.py [-h] [-v] [-c FILE]
                {status,query,hostadd,alias,hostdel,hostlist,hostsearch,zoneadd,zonedel,zonelist,zonestatus}
                ...

mmbop manages BIND over Python

optional arguments:
  -h, --help                      show this help message and exit
  -v, --verbose                   Enable verbose messages
  -c FILE, --config FILE          Location of config file

commands:
  DNS actions

  {status,query,hostadd,alias,hostdel,hostlist,hostsearch,zoneadd,zonedel,zonelist,zonestatus}
                                  add -h after command for additional
                                  information
    status                        Return status of BIND
    query                         Query for name or IP
    hostadd                       Add an A and PTR record
    alias                         Add a CNAME record
    hostdel                       Remove CNAME or A and PTR record
    hostlist                      Show all records for zone
    hostsearch                    Wildcard search of a zone
    zoneadd                       Add a zone
    zonedel                       Remove a zone
    zonelist                      Show all zones
    zonestatus                    Show status of a zone
```

Note that the *zonelist* option will only show the domains that mmbop can manage (using same validation criteria as for adding/removing zones) - see above for the protect/require configuration. Also note that list works by running 'rndc dumpdb -zones' and then parses this file to obtain the applicable domains. It has to wait for the dump file to complete writing, which can take a surprisingly long (relatively speaking) time - 6+ seconds if you have a lot of large zone files.

## API

It may be desirable to provide the ability to add/remove zones, without wanting to give these users shell access to the primary DNS server. The script *mmbop.py* provides a simple REST API interface, which can be used with your favorite [WSGI](https://www.python.org/dev/peps/pep-3333/) capable web server to provide remote access to the application. There is a simple header-based authorization token solution for validating requests; for production use you will likely want something more robust (and/or handled directly through the web server controls).

**Step 1 : Configure mmbop_api.ini**

The only item in this configuration file is the SHA224 hashed hexidecimal digest of the token string that the client will provide in the web request.

To generate the hash, you can run python directly:

```
$ python3 -q
>>> import hashlib
>>> clear_token = 'mysupersecretkey'
>>> hash_token = hashlib.sha224(clear_token.encode()).hexdigest()
>>> print(hash_token)
fb096d8fa48b12cf2adec03e5e5d03fb231bb87674f8d8dbf137f05c
```
Added to the *mmbop_api.ini* file:

```
[DEFAULT]
token:fb096d8fa48b12cf2adec03e5e5d03fb231bb87674f8d8dbf137f05c
```
Given this, for a request to be valid the client must send *mysupersecretkey* as the value in the *Authorization* field of the header. See the examples below.

**Step 2 : Start service**

For the purpose of providing an example, gunicorn will be directly used. By default this starts the service listening locally on port 8000.

See [this guide](https://www.digitalocean.com/community/tutorials/how-to-deploy-falcon-web-applications-with-gunicorn-and-nginx-on-ubuntu-16-04) for using gunicorn with nginx.

For production, you should use https to keep the auth token secure.

```
$ sudo gunicorn mmbop_api:APP
[2019-08-22 15:35:53 -0400] [16586] [INFO] Starting gunicorn 19.9.0
[2019-08-22 15:35:53 -0400] [16586] [INFO] Listening at: http://127.0.0.1:8000 (16586)
[2019-08-22 15:35:53 -0400] [16586] [INFO] Using worker: sync
[2019-08-22 15:35:53 -0400] [16589] [INFO] Booting worker with pid: 16589
```

Example 1: Listing all zones:

```
$ curl -i -H "Content-Type: application/json" -H "Authorization: mysupersecretkey" http://127.0.0.1:8000/zonelist
HTTP/1.1 200 OK
Server: gunicorn/19.9.0
Date: Thu, 22 Aug 2019 19:42:08 GMT
Connection: close
content-length: 41
content-type: application/json
scott.example.com sarah.example.net
```
The use of *-i* in curl is just to show the response code, not required for obtaining a result

Example 2: Adding an invalid zone (does not meet the protect/require standards):

```
$ curl -i -X POST -H "Content-Type: application/json" -H "Authorization: mysupersecretkey" -d '{"domain":"bogus.company.com"}' http://127.0.0.1:8000/modify
HTTP/1.1 400 Bad Request
Server: gunicorn/19.9.0
Date: Thu, 22 Aug 2019 20:14:45 GMT
Connection: close
content-length: 43
content-type: application/json

Not a valid zone name. Check configuration.
```

**Current list of API functions and their allowed methods:**

- /query
    - equivalent to: mmbop query <entry>
    - GET
    - Required params: entry
- /status
    - equivalent to: mmbop status
    - GET
- /hostmodify
    - equivalent to: mmbop hostadd <fqdn> <add> or mmbop hostdel <fqdn|addr>
    - POST (adding entries), DELETE (removing entries)
    - Required params: fqdn, addr (if POST)
    - Optional params: force
- /alias
    - equivalent to: mmbop alias <alias> <real>
    - POST (adding entries; for removing, use /hostmodify)
    - Required params: alias, real
    - Optional params: force
- /hostlist/{domain}
    - equivalent to: mmbop <domain>
    - GET
- /hostsearch
    - equivalent to: mmbop <domain> <term>
    - GET
    - Required params: domain
    - Optional params: term
- /zonemodify
    - equivalent to: mmbop zoneadd <domain> or mmbop zonedel <domain>
    - POST (adding zones), DELETE (removing zones)
    - Required params: domain
- /zonelist
    - equivalent to: mmbop zonelist
    - GET
- /zoneinfo/{domain}
    - equivalent to: mmbop zonestatus <domain>
    - GET

## Built With

* [VIM](https://www.vim.org/) - venerable and more than capable
* [Pylint](https://www.pylint.org/) - keeping my code somewhat under control and consistent

## Authors

* **Scott Strattner** - [IBM Github](https://github.ibm.com/sstrattn)

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* Thanks to [dnspython](http://www.dnspython.org/) I was able to reverse engineer the weird 'wire format' required for catalog zone entries
* [This](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2) nice README template


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Initial Release

1.0 - 2020-01-16

### Added
- Added Changelog and 


mmbop_api.py
"""
Uses Falcon to implement simple API wrapper for mmbop
"""
import ipaddress
import configparser
import hashlib
import json
import falcon
import mmbop

# Pylint really doesn't like falcon, this is to silence false positives
# pylint: disable=too-few-public-methods,c-extension-no-member,no-self-use,no-member

class HandleCORS:
    """
    To enable all sites to reach the API, as it relies on
    token authentication for access control
    """

    def process_request(self, req, resp):
        """
        Taken from:
        https://github.com/falconry/falcon/issues/1220
        """
        resp.set_header('Access-Control-Allow-Origin', '*')
        resp.set_header('Access-Control-Allow-Methods', '*')
        resp.set_header('Access-Control-Allow-Headers', '*')
        resp.set_header('Access-Control-Max-Age', 1728000)  # 20 days
        if req.method == 'OPTIONS':
            raise falcon.http_status.HTTPStatus(falcon.HTTP_200, body='\n')

class AuthToken:
    """
    Implements simple authentication token
    """

    # Contains the string (hexdigest) of the SHA224 hashing of the token
    AUTH_TOKEN = None
    # Location where hashed token string can be found
    TOKEN_FILE = './mmbop_api.ini'
    # Key in the token file, the value of which is the hashed token string
    TOKEN_FIELD = 'token'

    def __init__(self):
        if not self.__class__.AUTH_TOKEN:
            self.get_token()
        if not self.__class__.AUTH_TOKEN:
            raise falcon.HTTPError(falcon.HTTP_500, 'Auth Error', 'Unable to find auth token')

    def get_token(self):
        """
        Attempt to read token file and fill AUTH_TOKEN
        """
        config = configparser.ConfigParser()
        try:
            config.read(self.__class__.TOKEN_FILE)
            default_config = config['DEFAULT']
            self.__class__.AUTH_TOKEN = default_config.get(self.__class__.TOKEN_FIELD, None)
        except (IOError, configparser.Error):
            pass

    def process_request(self, req, _):
        """
        Falcon required method for handling request and (in this case)
        ensuring the provided token is correct
        """
        token = req.get_header('Authorization')
        if not token:
            desc = ('Provide an authentication token in header request')
            raise falcon.HTTPUnauthorized('Auth token required', desc)
        if not self._token_is_valid(token):
            desc = ('Invalid authentication token')
            raise falcon.HTTPUnauthorized('Invalid auth token', desc)

    def _token_is_valid(self, token):
        return bool(self._hash(token) == self.__class__.AUTH_TOKEN)

    @staticmethod
    def _hash(token):
        """
        Perform SHA224 hash on given token and return string (hexdigest)
        """
        return hashlib.sha224(str.encode(token)).hexdigest()

class RNDCBase:
    """
    Base API class to initialize rndc
    """

    def __init__(self, rndc_instance):
        self.rndc = rndc_instance

class NSBase:
    """
    Base API class to initialize nsupdate
    """

    def __init__(self, nsupdate_instance):
        self.nsupdate = nsupdate_instance

class DIGBase:
    """
    Base API class to initialize dig
    """

    def __init__(self, dig_instance):
        self.dig = dig_instance

class Query(DIGBase):
    """
    Handle query requests
    """

    def on_get(self, req, resp):
        """
        Equivalent to "mmbop query <entry>" command
        """
        entry = req.params.get('entry', None)
        reply = []
        matched_entries = []
        if entry:
            matched_entries = self.dig.find_record(entry)
        if not matched_entries:
            reply.append({'entry': 'Record not found'})
        for match in matched_entries:
            try:
                (host, ttl, _, class_type, value) = match.split()
                reply.append({'entry': host, 'ttl': ttl,
                              'class': class_type, 'value': value})
            except ValueError:
                continue
        resp.body = json.dumps({'matched_entries': reply})

class Status(RNDCBase):
    """
    Handle status requests
    """

    def on_get(self, _, resp):
        """
        Return result of 'rndc status' on GET request
        Reply json:
            { 'server_status': [{ 'line': <line1>},
                                { 'line': <line2>},
                                ...
                               ]
            }
        """
        server_status = self.rndc.status()
        response_lines = [{'line': x} for x in server_status.split('\n')]
        resp.body = json.dumps({'server_status': response_lines})

class AliasAdd(NSBase):
    """
    Handle alias requests

    Returns JSON of form:
    {'results': [{ 'success': boolean, 'message': string }]}

    HTTP response code will be 201 (on adds) if result succeeded.
    Otherwise, response code will be 400.
    """

    def on_post(self, req, resp):
        """
        Equivalent to "mmbop.py alias <alias> <real>" command
        """
        result_list = []
        alias = req.media.get('alias')
        real = req.media.get('real')
        force = req.media.get('force', False)
        if not (alias and real):
            result_list.append({'success': False,
                                'message': 'Need to provide the alias and real name'})
        (success, message) = self.nsupdate.add_alias(alias, real, force)
        if success:
            message = alias + ' ' + real + ' added successfully'
            result_list.append({'success': True, 'message': message})
        else:
            message = alias + ' ' + real + ' not added: ' + message
            result_list.append({'success': False, 'message': message})
        if result_list[0]['success']:
            resp.status = falcon.HTTP_201
        else:
            resp.status = falcon.HTTP_400
        resp.body = json.dumps({'result': result_list})

class HostModify(NSBase):
    """
    Handle hostmodify requests

    Returns JSON of form:
    {'results': [{ 'success': boolean, 'message': string },
                 { 'success': boolean, 'message': string },
                 ...
                ]
    HTTP response code will be 201 (on adds) or 200 (on deletes)
    if all results succeeded. Otherwise, will be 400.
    """

    @staticmethod
    def _all_succeeded(results, delete=False):
        """
        Sets the response status code. If all transactions were
        successful, the response code will be 201 (for add) or
        200 (for delete). Otherwise, response code will be 400.
        """
        for result in results:
            if not result['success']:
                return falcon.HTTP_400
        if delete:
            return falcon.HTTP_200
        return falcon.HTTP_201

    def on_post(self, req, resp):
        """
        Equivalent to "mmbop.py hostadd <fqdn> <ip_addr> ..." command
        """
        result_list = []
        fqdn = req.media.get('fqdn')
        addr = req.media.get('addr')
        force = req.media.get('force', False)
        if not (fqdn and addr):
            result_list.append({'success': False,
                                'message': 'Need to provide the fqdn and addr'})
        else:
            if ' ' in addr:
                addr = addr.split(' ')
            (success, message) = self.nsupdate.add_record(fqdn, addr, force)
            if success:
                message = fqdn + ' ' + str(addr) + ' added successfully'
                result_list.append({'success': True, 'message': message})
            else:
                message = fqdn + ' ' + str(addr) + ' not added: ' + message
                result_list.append({'success': False, 'message': message})
        resp.status = self._all_succeeded(result_list)
        resp.body = json.dumps({'result': result_list})

    def on_delete(self, req, resp):
        """
        Equivalent to "mmbop.py hostdel <entry>" command
        """
        result_list = []
        entry = req.media.get('entry')
        force = req.media.get('force', False)
        if not entry:
            result_list.append({'success': False,
                                'message': 'Need to provide the fqdn or address'})
        else:
            (success, message) = self.nsupdate.delete_record(entry, force)
            if success:
                message = entry + ' deleted successfully'
                result_list.append({'success': True, 'message': message})
            else:
                message = entry + ' not deleted: ' + message
                result_list.append({'success': False, 'message': message})
        resp.status = self._all_succeeded(result_list)
        resp.body = json.dumps({'result': result_list})

class HostList(DIGBase):
    """
    Handle hostlist requests
    """

    def on_get(self, _, resp, domain):
        """
        Equivalent to 'mmbop.py hostlist <domain>'
        """
        resp.body = (self.dig.hostlist(domain))

class HostSearch(DIGBase):
    """
    Handle hostsearch requests
    """

    def on_get(self, req, resp):
        """
        Equivalent to 'mmbop.py hostlist <domain> <term>'.
        Call with query parameters: ?domain=xxx&term=yyy.
        Prepending '~' to the term will enable negation -
        returning all entries that do not match the term.
        Reply json:
            {matched_entries: [{entry: <entry1>, value: <value1>},
                               {entry: <entry2>, value: <value2>},
                               ...
                              ]
            }
        """
        reverse = False
        domain = self.fix_if_reverse(req.params.get('domain', None))
        term = self.reverse_if_ip(domain, req.params.get('term', None))
        search_term = term
        matched_entries = []
        reply = []
        if term and term.startswith('~'):
            reverse = True
            search_term = term[1:]
        if domain:
            matched_entries = self.dig.search_domain(domain, search_term, reverse)
        if not matched_entries:
            message = 'No records found'
            if reverse:
                message = message + ' not'
            if term:
                message = message + ' containing term "' + search_term + '"'
            reply.append({'entry': message, 'value': None})
            matched_entries = [message]
        for match in matched_entries:
            try:
                (entry, _, _, _, value) = match.split()
                reply.append({'entry': entry, 'value': value})
            except ValueError:
                continue
        resp.body = json.dumps({'matched_entries': reply})

    @staticmethod
    def fix_if_reverse(zone_name):
        """
        If a user requests a search of a reverse zone, but
        gives it as (for example) 192.168.1 instead of
        1.168.192.in-addr.arpa, convert it to proper reverse
        zone name format
        """
        zone_entries = zone_name.split('.', 3)
        for zone_slice in zone_entries[:3]:
            if not zone_slice.isdigit():
                return zone_name
        if 'in-addr.arpa' not in zone_entries:
            new_zone = zone_entries[2] + '.' + zone_entries[1] + '.' + zone_entries[0]
            new_zone += '.in-addr.arpa'
            return new_zone
        return zone_name

    @staticmethod
    def reverse_if_ip(domain, search_term):
        """
        If the user is searching a reverse zone (*.in-addr.arpa),
        and the search term is an IP address, reverse it so that
        it can be found.
        """
        reverse_domain_tag = '.in-addr.arpa'
        if reverse_domain_tag in domain:
            try:
                search_is_ip = ipaddress.ip_address(search_term).reverse_pointer
                return search_is_ip.replace(reverse_domain_tag, '')
            except ValueError:
                pass
        return search_term

class ZoneInfo(RNDCBase):
    """
    Handle status request for specific domain
    """

    def on_get(self, _, resp, domain):
        """
        Return result of 'rndc zoneinfo <zone>' on GET request
        """
        resp.body = (self.rndc.zonestatus(domain))

class ZoneList(RNDCBase):
    """
    Handle list of active zones
    """

    def on_get(self, _, resp):
        """
        Return result of 'rndc dumpdb' with zone checking
        Reply json:
            {zones: [{zone: <zone1>},
                     {zone: <zone2>},
                     ...
            }
        """
        zones = []
        zone_list = self.rndc.list_zones()
        for zone in zone_list:
            zones.append({'zone': zone})
        resp.body = json.dumps({'zones': zones})

class ZoneModify(RNDCBase):
    """
    Handle add/delete requests
    """

    def on_post(self, req, resp):
        """
        Adding a zone via POST request
        """
        zone_name = req.media.get('domain')
        if not zone_name:
            resp.status = falcon.HTTP_400
            resp.body = ('Missing domain parameter')
        else:
            (success, message) = self.rndc.add(zone_name)
            if success:
                resp.status = falcon.HTTP_201
                resp.body = json.dumps({'result': [{'success': True, 'message': message}]})
            else:
                resp.status = falcon.HTTP_400
                resp.body = json.dumps({'result': [{'success': False, 'message': message}]})

    def on_delete(self, req, resp):
        """
        Removing a zone via DELETE request
        """
        zone_name = req.media.get('domain')
        if not zone_name:
            resp.statuss = falcon.HTTP_400
            resp.body = ('Missing domain parameter')
        else:
            (success, message) = self.rndc.delete(zone_name)
            if success:
                resp.body = json.dumps({'result': [{'success': True, 'message': message}]})
            else:
                resp.status = falcon.HTTP_400
                resp.body = json.dumps({'result': [{'success': True, 'message': message}]})


CONF_FILE = './mmbop.ini'

MY_CONF = mmbop.read_config(CONF_FILE)
MY_RNDC = mmbop.RNDC.create(**MY_CONF)
MY_NSUPDATE = MY_RNDC.my_nsupdate
MY_DIG = mmbop.DigQuery()
APP = falcon.API(middleware=[HandleCORS(), AuthToken()])
QUERY = Query(MY_DIG)
STATUS = Status(MY_RNDC)
HOSTMODIFY = HostModify(MY_NSUPDATE)
ALIASADD = AliasAdd(MY_NSUPDATE)
HOSTLIST = HostList(MY_DIG)
HOSTSEARCH = HostSearch(MY_DIG)
ZONEINFO = ZoneInfo(MY_RNDC)
ZONEMODIFY = ZoneModify(MY_RNDC)
ZONELIST = ZoneList(MY_RNDC)
APP.add_route('/query', QUERY)
APP.add_route('/status', STATUS)
APP.add_route('/hostmodify', HOSTMODIFY)
APP.add_route('/alias', ALIASADD)
APP.add_route('/hostlist/{domain}', HOSTLIST)
APP.add_route('/hostsearch', HOSTSEARCH)
APP.add_route('/zonemodify', ZONEMODIFY)
APP.add_route('/zonelist', ZONELIST)
APP.add_route('/zoneinfo/{domain}', ZONEINFO)
